File: Serenity.Script.UI/FilterPanel/FilteringTypes/StringFiltering.cs
using System.Collections.Generic;
using System.Runtime.CompilerServices;

namespace Serenity
{
    [Imported(ObeysTypeSystem = true)]
    public class StringFiltering : BaseFiltering
    {
        public override List<FilterOperator> GetOperators()
        {
            return AppendNullableOperators(new List<FilterOperator>
            {
                FilterOperators.Contains,
                FilterOperators.StartsWith,
                FilterOperators.EQ,
                FilterOperators.NE,
                FilterOperators.BW
            });
        }

        protected override object ValidateEditorValue(string value)
        {
            if (value.Length == 0)
                return value;

            return base.ValidateEditorValue(value);
        }
    }
}

File: packages/corelib/src/ui/filtering/stringfiltering.ts
import { nsSerenity } from "../../base";
import { BaseFiltering } from "./basefiltering";
import { FilterOperator, FilterOperators } from "./filteroperator";

/**
 * Filtering handler for string fields.
 */
export class StringFiltering extends BaseFiltering {
    static override[Symbol.typeInfo] = this.registerClass(nsSerenity);

    /**
     * Returns the operators supported by this filtering handler.
     * @returns The operators.
     */
    getOperators(): FilterOperator[] {
        var ops = [
            { key: FilterOperators.contains },
            { key: FilterOperators.startsWith },
            { key: FilterOperators.EQ },
            { key: FilterOperators.NE }
        ];
        return this.appendNullableOperators(ops);
    }

    /**
     * Validates the editor value, allowing empty values.
     * @param value - The value to validate.
     * @returns The validated value.
     */
    override validateEditorValue(value: string) {
        if (value.length === 0) {
            return value;
        }

        return super.validateEditorValue(value);
    }
}
